styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java

```diff
@@ -91,72 +104,110 @@ void tick() {
+                initialNextTrigger, remainingCapacity))) {
+          break;
+        }
+      } catch (IOException e) {
+        // transaction failure, yield
+        LOG.debug("Transaction failure when working on backfill {}", backfill, e);
+        break;
       }
+    }
+  }
 
-      partition = nextInstant(partition, backfill.schedule());
-      storeBackfill(backfill.builder()
-                        .nextTrigger(partition)
-                        .build());
+  private boolean triggerNextPartitionAndProgress(StorageTransaction tx,
+                                                  String id,
+                                                  Workflow workflow,
+                                                  Instant initialNextTrigger,
+                                                  int remainingCapacity) {
+    final Backfill momentBackfill = tx.backfill(id).orElseThrow(RuntimeException::new);
+    final Instant momentNextTrigger = momentBackfill.nextTrigger();
+
+    if (numberOfInstants(momentNextTrigger, initialNextTrigger,
+        momentBackfill.schedule()) >= remainingCapacity) {
+      LOG.debug("Capacity reached for backfill {}", momentBackfill);
+      return false;
     }
 
-    if (partition.equals(backfill.end())) {
-      storeBackfill(backfill.builder()
-                        .nextTrigger(backfill.end())
-                        .allTriggered(true)
-                        .build());
+    if (momentNextTrigger.equals(momentBackfill.end())) {
+      LOG.debug("Backfill {} all triggered", momentBackfill);
+      return false;
     }
-  }
 
-  private Map<String, Long> getBackfillStates() {
-    final List<InstanceState> activeStates = stateManager.activeStates().entrySet().stream()
-        .map(entry -> InstanceState.create(entry.getKey(), entry.getValue()))
-        .collect(toList());
-
-    return activeStates.stream()
-        .map(state -> state.runState().data().trigger())
-        .filter(Optional::isPresent)
-        .map(Optional::get)
-        .filter(TriggerUtil::isBackfill)
-        .collect(groupingBy(
-            TriggerUtil::triggerId,
-            HashMap::new,
-            counting()));
+    try {
+      // race conditions:
+      //   - the other scheduler instance has triggered partition, but hasn't committed the
+      //     transaction, and the workflow instance has finished very quickly
+      //   - this scheduler reads initialNextTrigger before the other scheduler commits the
+      //     transaction, so it will move on to trigger the same partition again, but we still
+      //     don't violate concurrency limit in this case
+      final CompletableFuture<Void> processed = triggerListener
+          .event(workflow, Trigger.backfill(momentBackfill.id()), momentNextTrigger)
+          .toCompletableFuture();
+      // Wait for the trigger execution to complete before proceeding to the next partition
+      processed.get();
```

Comment:
It seems one could just `.event().wait()` instead of going `toCompletableFuture()` in between, but I guess this was preexisting code...

Reply:
Shouldn't we have a timeout? `get()` blocks indefinitely...

Reply:
I believe we'd have this problem in many other places in the code. I suggest to open a new PR about this in the future if we want to tackle this.